Mathlib/CategoryTheory/Topos/Classifier.lean

```diff
@@ -48,68 +49,126 @@ Let `C` refer to a category with a terminal object.
+  Ω₀ : C
+  /-- The codomain of the truth morphism -/
   Ω : C
-  /-- the truth morphism for a subobject classifier -/
-  truth : ⊤_ C ⟶ Ω
+  /-- The truth morphism of the subobject classifier -/
+  truth : Ω₀ ⟶ Ω
+  /-- The truth morphism is a monomorphism -/
+  mono_truth : Mono truth
+  /-- The top arrow in the pullback square -/
+  χ₀ {U X : C} (m : U ⟶ X) [Mono m] : U ⟶ Ω₀
   /-- For any monomorphism `U ⟶ X`, there is an associated characteristic map `X ⟶ Ω`. -/
   χ {U X : C} (m : U ⟶ X) [Mono m] : X ⟶ Ω
-  /-- `χ m` forms the appropriate pullback square. -/
-  isPullback {U X : C} (m : U ⟶ X) [Mono m] : IsPullback m (terminal.from U) (χ m) truth
-  /-- `χ m` is the only map `X ⟶ Ω` which forms the appropriate pullback square. -/
-  uniq {U X : C} (m : U ⟶ X) [Mono m] (χ' : X ⟶ Ω)
-      (hχ' : IsPullback m (terminal.from U) χ' truth) :
-    χ' = χ m
+  /-- `χ m` and `χ₀ m` form the appropriate pullback square. -/
+  isPullback' {U X : C} (m : U ⟶ X) [Mono m] : IsPullback m (χ₀ m) (χ m) truth
+  /-- `χ m` is the only map `X ⟶ Ω` which forms the appropriate pullback square for any `χ₀'`. -/
+  uniq' {U X : C} (m : U ⟶ X) [Mono m] (χ₀' : U ⟶ Ω₀) (χ' : X ⟶ Ω)
+    (hχ' : IsPullback m χ₀' χ' truth) : χ' = χ m
+
+variable {C}
+
+namespace Classifier
+
+attribute [instance] mono_truth
+
+/-- More explicit constructor in case `Ω₀` is already known to be a terminal object. -/
+def fromTerminalΩ₀
+    (Ω₀ : C)
+    (t : IsTerminal Ω₀)
+    (Ω : C)
+    (truth : Ω₀ ⟶ Ω)
+    (χ : ∀ {U X : C} (m : U ⟶ X) [Mono m], X ⟶ Ω)
+    (isPullback : ∀ {U X : C} (m : U ⟶ X) [Mono m],
+      IsPullback m (t.from U) (χ m) truth)
+    (uniq : ∀ {U X : C} (m : U ⟶ X) [Mono m] (χ' : X ⟶ Ω)
+      (_ : IsPullback m (t.from U) χ' truth), χ' = χ m) : Classifier C where
+  Ω₀ := Ω₀
+  Ω := Ω
+  truth := truth
+  mono_truth := t.mono_from _
+  χ₀ _ _ := t.from _
+  χ m _ := χ m
+  isPullback' m _ := isPullback m
+  uniq' m _ χ₀' χ' hχ' := uniq m χ' ((t.hom_ext χ₀' (t.from _)) ▸ hχ')
+
+/-- `Ω₀` is a terminal object. -/
+def isTerminalΩ₀ {c : Classifier C} : IsTerminal c.Ω₀ :=
+    have : ∀ Y : C, Unique (Y ⟶ c.Ω₀) := fun Y =>
+      { default := c.χ₀ (𝟙 Y),
+        uniq f :=
+          have : f ≫ c.truth = c.χ₀ (𝟙 Y) ≫ c.truth :=
+            by calc
+              _ = c.χ (𝟙 Y) := c.uniq' (𝟙 Y) f (f ≫ c.truth) (of_horiz_isIso_mono (by simp))
+              _ = c.χ₀ (𝟙 Y) ≫ c.truth := by simp [← (c.isPullback' (𝟙 Y)).w]
+          Mono.right_cancellation _ _ this }
+    IsTerminal.ofUnique c.Ω₀
+
+/-- The more practical version of `isPullback'` where we don't need to supply the to arrow,
+since there is a unique morphism to the terminal object -/
+lemma isPullback {U X : C} {c : Classifier C} (m : U ⟶ X) [Mono m] :
+    IsPullback m (isTerminalΩ₀.from _) (c.χ m) c.truth :=
+  (isTerminalΩ₀.hom_ext (c.χ₀ m) (isTerminalΩ₀.from U)) ▸ c.isPullback' m
+
+/-- The unique morphism from `U` to the terminal object `Ω₀` -/
+def uniqueToΩ₀ {c : Classifier C} (U : C) : U ⟶ c.Ω₀ :=
```

Comment:
```suggestion
def toΩ₀ {c : Classifier C} (U : C) : U ⟶ c.Ω₀ :=
```
maybe uniqueness does not have to be part of the name and would make things shorter to write?
The name of `uniqueToΩ₀_eq_χ₀_id` would need to be adapted.